src/main/java/jetbrains/buildServer/auth/oauth/ServerPrincipalFactory.java

```diff
@@ -18,36 +20,108 @@
+            boolean allowCreatingNewUsersByLogin) {
+        Optional<SUser> existingUserOptional = findExistingUser(user.getId());
+        boolean syncGroups = properties.isSyncGroups();
+        List<String> groupsInToken = user.getGroups();
+        if (existingUserOptional.isPresent()) {
             LOG.info("Use existing user: " + user.getId());
-            return existingPrincipal;
+            SUser existingUser = existingUserOptional.get();
+            if (syncGroups) {
+                //user group or all user groups?
+                List<String> existingUserGroups =
+                        existingUser.getUserGroups().stream().map(userGroup -> userGroup.getName())
+                                .collect(Collectors.toList());
+                List<String> additionalGroupsInToken = obtainGroupsDelta(groupsInToken, existingUserGroups);
+                addUserToGroups(existingUser, applyGroupWhitelist(additionalGroupsInToken));
+
+                List<String> groupsNotPresentInToken = obtainGroupsDelta(existingUserGroups, groupsInToken);
+                removeUserFromGroups(existingUser, applyGroupWhitelist(groupsNotPresentInToken));
+            }
+            return Optional.of(new ServerPrincipal(PluginConstants.OAUTH_AUTH_SCHEME_NAME, existingUser.getUsername()));
         } else if (allowCreatingNewUsersByLogin) {
             LOG.info("Creating user: " + user);
             SUser created = userModel.createUserAccount(PluginConstants.OAUTH_AUTH_SCHEME_NAME, user.getId());
+            if (syncGroups) {
+                addUserToGroups(created, applyGroupWhitelist(groupsInToken));
+            }
             created.setUserProperty(PluginConstants.ID_USER_PROPERTY_KEY, user.getId());
             created.updateUserAccount(user.getId(), user.getName(), user.getEmail());
             return Optional.of(new ServerPrincipal(PluginConstants.OAUTH_AUTH_SCHEME_NAME, user.getId()));
         } else {
             LOG.info("User: " + user + " could not be found and allowCreatingNewUsersByLogin is disabled");
-            return existingPrincipal;
+            return Optional.empty();
+        }
+    }
+
+    private void removeUserFromGroups(SUser existingUser, List<String> groupsToDelete) {
+        for (String group : groupsToDelete) {
+            SUserGroup userGroup = userGroupManager.findUserGroupByName(group);
+            if (userGroup != null) {
+                userGroup.removeUser(existingUser);
+            }
+        }
+    }
+
+    private List<String> applyGroupWhitelist(List<String> groups) {
+        List<String> finalGroupList = new ArrayList<>();
+        Set<String> whitelistedGroups = properties.getWhitelistedGroups();
+        if (whitelistedGroups != null) {
+            for (String group : groups) {
+                for (String whitelistedGroup : whitelistedGroups) {
+                    if (group.startsWith(whitelistedGroup)) {
```

Comment:
why do we check startsWith instead of equals

Reply:
The intention is to provide limited support for regex with startsWith. The UI plugin has the following label for this field `Groups (starts with):` and `ex: "dev-team1, dev-team2" or "dev-"`